modules/memfault-firmware-sdk/memfault_location_metrics.c

```diff
@@ -0,0 +1,209 @@
+					       enum location_event_id id,
+					       const struct location_data_details *details,
+					       const float *accuracy_m)
+{
+	switch (method) {
+#if defined(CONFIG_LOCATION_METHOD_GNSS)
+	case LOCATION_METHOD_GNSS:
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_method_time_ms,
+					     details->elapsed_time_method);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_method_result, id);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_on_time_ms,
+					     details->gnss.elapsed_time_gnss);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_satellites_tracked_count,
+					     details->gnss.satellites_tracked);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_satellites_used_count,
+					     details->gnss.satellites_used);
+
+		/* Only record TTF and accuracy when a fix is acquired */
+		if (id == LOCATION_EVT_LOCATION) {
+			MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_time_to_fix_ms,
+						     details->gnss.pvt_data.execution_time);
+
+			if (accuracy_m != NULL) {
+				uint32_t accuracy_cm = (uint32_t)(*accuracy_m * 100.0f);
+
+				MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_accuracy_cm, accuracy_cm);
+			}
+		}
+		break;
+#endif
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR)
+	case LOCATION_METHOD_CELLULAR:
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_lte_method_time_ms,
+					     details->elapsed_time_method);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_lte_method_result, id);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_lte_neighbor_cells_count,
+					     details->cellular.ncells_count);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_lte_gci_cells_count,
+					     details->cellular.gci_cells_count);
+		break;
+#endif
+#if defined(CONFIG_LOCATION_METHOD_WIFI)
+	case LOCATION_METHOD_WIFI:
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_wifi_method_time_ms,
+					     details->elapsed_time_method);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_wifi_method_result, id);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_wifi_ap_count,
+					     details->wifi.ap_count);
+		break;
+#endif
+	default:
+		LOG_DBG("Unsupported location method id=%d", method);
+		break;
+	}
+}
+
+/**
+ * @brief Record the fallback method used
+ *
+ * @param method fallback method
+ * @param fallback pointer to fallback method data
+ */
+static void prv_location_method_fallback_record(enum location_method method,
+						const struct location_data_fallback *fallback)
+{
+	prv_location_method_results_record(method, fallback->cause, &fallback->details, NULL);
+
+	/* Log the results */
+	char method_str[16];
+
+	snprintk(method_str, sizeof(method_str), "%s", prv_location_method_to_string(method));
+	LOG_DBG("%s method attempted", method_str);
+
+	char fallback_method_str[16];
+
+	snprintk(fallback_method_str, sizeof(fallback_method_str), "%s",
+		 prv_location_method_to_string(fallback->next_method));
+	LOG_DBG("Falling back to %s method", fallback_method_str);
+}
+
+/**
+ * @brief Start a location session.
+ *
+ */
+static void prv_location_metrics_session_start(void)
+{
+	if (s_session_in_progress) {
+		return;
```

Comment:
I'd add `LOG_WRN` etc. here (and wherever in progress is checked) to show that operation was blocked.